Cluster unknown faces alongside known faces

jakekapellen opened the pull request:
This is a significant enhancement to the clustering algorithm. Now, it clusters unknown faces too when you select the Train Faces button. It even estimates the probability that each unknown faces corresponds to an unknown cluster.

jakekapellen committed:
Initial commit to cluster unknown faces

jakekapellen committed:
Merge branch 'LibrePhotos:dev' into dev

jakekapellen committed:
Redesign of new clustering algorithm

jakekapellen committed:
Significant redesign and fixing of probability calculations

jakekapellen committed:
Delete lots of unused code

jakekapellen committed:
Minor cleanup

jakekapellen committed:
Change how the code finds the "unknown" person

jakekapellen committed:
Minor coding style tweak

jakekapellen committed:
Auto-create unknown person if it doesn't exist

jakekapellen committed:
Add default to allow migration to be reverted

jakekapellen committed:
Documentation for key clustering functions

jakekapellen committed:
Auto-retrain faces after whole-folder scans

derneuere commented:
Works well and the code looks good! Very awesome :+1:

You only forgot to use the pre commit hook, I think, and I encountered one issue:

How to recreate:
Scan and cluster with your first user. Create another account, scan some other images, and then try to cluster them. The job will immediately stop. If you go to the admin panel, you can see that the job failed. 
You will also find this error message within the logs:

```
2022-07-18 19:53:29,674 : api_util.py : get_search_term_examples : 195 : INFO : Found 100 photos
2022-07-18 19:53:37,833 : face_classify.py : cluster_all_faces : 98 : ERROR : An error occurred
Traceback (most recent call last):
  File "/code/api/face_classify.py", line 83, in cluster_all_faces
    delete_clusters()
  File "/code/api/face_classify.py", line 159, in delete_clusters
    ClusterManager.delete_cluster(cluster)
  File "/code/api/cluster_manager.py", line 29, in delete_cluster
    unknown_cluster: Cluster = ClusterManager.get_unknown_cluster()
  File "/code/api/cluster_manager.py", line 132, in get_unknown_cluster
    unknown_cluster: Cluster = Cluster.get_or_create_cluster_by_id(UNKNOWN_CLUSTER_ID)
  File "/code/api/models/cluster.py", line 30, in get_or_create_cluster_by_id
    return Cluster.objects.get_or_create(cluster_id=cluster_id)[0]
  File "/usr/local/lib/python3.9/dist-packages/django/db/models/manager.py", line 85, in manager_method
    return getattr(self.get_queryset(), name)(*args, **kwargs)
  File "/usr/local/lib/python3.9/dist-packages/django/db/models/query.py", line 573, in get_or_create
    return self.get(**kwargs), False
  File "/usr/local/lib/python3.9/dist-packages/django/db/models/query.py", line 433, in get
    raise self.model.MultipleObjectsReturned(
api.models.cluster.Cluster.MultipleObjectsReturned: get() returned more than one Cluster -- it returned 2!
2022-07-18 19:54:58,264 : face_classify.py : cluster_all_faces : 98 : ERROR : An error occurred
Traceback (most recent call last):
  File "/code/api/face_classify.py", line 83, in cluster_all_faces
    delete_clusters()
  File "/code/api/face_classify.py", line 159, in delete_clusters
    ClusterManager.delete_cluster(cluster)
  File "/code/api/cluster_manager.py", line 29, in delete_cluster
    unknown_cluster: Cluster = ClusterManager.get_unknown_cluster()
  File "/code/api/cluster_manager.py", line 132, in get_unknown_cluster
    unknown_cluster: Cluster = Cluster.get_or_create_cluster_by_id(UNKNOWN_CLUSTER_ID)
  File "/code/api/models/cluster.py", line 30, in get_or_create_cluster_by_id
    return Cluster.objects.get_or_create(cluster_id=cluster_id)[0]
  File "/usr/local/lib/python3.9/dist-packages/django/db/models/manager.py", line 85, in manager_method
    return getattr(self.get_queryset(), name)(*args, **kwargs)
  File "/usr/local/lib/python3.9/dist-packages/django/db/models/query.py", line 573, in get_or_create
    return self.get(**kwargs), False
  File "/usr/local/lib/python3.9/dist-packages/django/db/models/query.py", line 433, in get
    raise self.model.MultipleObjectsReturned(
api.models.cluster.Cluster.MultipleObjectsReturned: get() returned more than one Cluster -- it returned 2!
2022-07-18 19:55:12,692 : api_util.py : get_search_term_examples : 181 : INFO : 100 possible ids
```
Probable cause:
I think it has something to do with clusters. You wrote in the issue, that an issue could be that clusters are not associated with users. I think specifically "Unknown - Other" or the numbering of the unknown clusters is off.

jakekapellen committed:
Fix pre-commit hook issues, clean up migrations

jakekapellen committed:
Fix remaining pre-commit issues

jakekapellen committed:
Sanity check when deleting unknown cluster

jakekapellen commented:
I believe I've addressed this - I cannot reproduce what you're describing, and I added a small sanity check when deleting the unknown cluster (to prevent it, actually).

Thanks for bearing with me; this is my first time using Python or pre-commit (we have other languages and processes where I work). I'm hoping this latest round of changes is good.

I am still not making clusters user-specific. This does cause one user's face training to "overwrite" another's. It doesn't seem like a massive deal, but I can add it now (or in a subsequent pull request), your choice.

jakekapellen committed:
Make clusters user-specific

jakekapellen commented:
I just went ahead and made clusters user-specific now anyway. It feels more usable now that different users can't accidentally trample each other's clusterings.

derneuere commented on api/models/cluster.py:
```diff
@@ -0,0 +1,40 @@
+
+    @staticmethod
+    def get_or_create_cluster_by_id(cluster_id: int):
+        return Cluster.objects.get_or_create(cluster_id=cluster_id)[0]
```
I solved my issue by wrapping this in a try except:
```
 @staticmethod
    def get_or_create_cluster_by_id(cluster_id: int):
		try:
			return Cluster.objects.get_or_create(cluster_id=cluster_id)[0]
		except:
			return Cluster.objects.filter(cluster_id=cluster_id).all()[0]
```
While this will definitely remove the error, it's still weird that an ID exists twice.

derneuere replied:
Haven't checked out your new solution yet! Will try that too.

jakekapellen replied:
Yeah, this would fix the crash for the exception, but I'd like to fix the underlying issue if possible. Either way, if it happens again we can wrap in try-except, but I think user-specific clusters should fix this.

jakekapellen committed:
Fix small bug with how we compute face probability

derneuere commented:
Now it works without the try except block! Awesome. 🚀 

I have encountered out of memory issues and long run times with a large dataset (>50k faces). It runs fine and is fast with 2k faces. I will run it again with the user specific clusters to see if I can reproduce it.

Usually, large queries without pagination can lead to OOM errors. Long run times, that only occur after a specific number of items, are usually related to quadratic time complexity somewhere in the code.

How large is your testing database?

jakekapellen commented:
The size depends on what I'm testing. I use 16-20 (very small) for full database rebuilds (so I can test migrations very quickly, for example), and only about 200 for larger clustering tasks.

I can try a larger dataset and see what happens. Sadly, my crappy old Macbook Air may not have room for a truly enormous data set, so I'll have to do my best.

It's also worth noting that DBSCAN is pretty performance-intensive. I considered moving to HDBSCAN, which should be faster. Alternatively, I could use OPTICS. I'll take note of the performance characteristics before I make either of those changes, though.

derneuere commented:
The runtime of DBSCAN according is in the worst case O(n^2) and in the best case O(n log(n)) [link](https://cs.stackexchange.com/questions/129545/what-would-be-the-time-complexity-of-dbscan-algorithm). I think this inefficiency gets introduced somewhere else, but that could be one reason :smile: 

You could use this dataset to get more faces: http://vis-www.cs.umass.edu/lfw/#download

jakekapellen committed:
Coding style and performance improvements

jakekapellen committed:
Small fix to not show unknown faces in albums

jakekapellen commented:
Thank you for sending over that dataset - that is very helpful!

I have solved the performance and memory issues. You were right that DBSCAN isn't the issue... The biggest changes I made with the latest commits were to not call bulk_update to apply the same changes repeatedly, and to rely on the built-in "on_deleted" properties in the model to clean up linked records. I was doing a lot of looping to clean up clusters and people, whereas I can just delete the fake "person" record, which in turn automatically cleans up the clusters and faces in one operation.

In short, calling something like "Face.objects.filter(...).update(...)" is **MUCH** faster than "for face in Face.object.filter(...): face.[field] = [value]" with a bulk_update.

Memory-wise, the background processes stay fixed around 340 MB. They start that way and don't seem to increase much when invoking DBSCAN. If you notice any more Out-of-Memory errors, please let me know and I'll address.

I took some performance numbers running over the 13K faces in the dataset you sent. Here's a breakdown of the numbers:
Total time: 188s (down from 567s initially)
     -Subtotal for [re-]clustering the faces: 52s 
          --Deleting clustered people: 1.7s
          --Deleting clusters: 1.7s
          --Creating new clusters subtotal: 49s (this was 380s initially)
               ---DBSCAN time: 1.7s
     -Subtotal for retraining faces and setting probabilities: 136s
          --Time for training the model based (work in MLPClassifier): 93.1s

derneuere commented:
Alright, I tested it with a very large dataset (170k faces) and did not find any OOM issues. It also finished in 20 minutes, which is in line with your findings, and it looks like it scales linearly. 
The frontend only shows 50k images anyway, so it should be future-proof! 😄  

Awesome work!
